### PR TITLE
Refactored Plugin.hs

### PR DESCRIPTION
**Pull request by bezirg**
The biggest change is the addition of an `type ECore = ExceptT CoreM` monad, so we can manage the errors ourselves.
Added some comments and tried to make the code more readable and shorter. Unfortunately I used implicitparams, because i think here it makes it cleaner.

### CODE REVIEW

File: plutus-core/plutus-ir/Language/PlutusIR/Compiler/Types.hs [hunk 1]
```diff
@@ -51,16 +51,16 @@ defaultCompilationOpts :: CompilationOpts
 defaultCompilationOpts = CompilationOpts True
 
 data CompilationCtx uni a = CompilationCtx {
-    _ccOpts        :: CompilationOpts
+    _ccOpts              :: CompilationOpts
     , _ccBuiltinMeanings :: PLC.DynamicBuiltinNameMeanings (PIR.Term PLC.TyName PLC.Name uni ())
-    , _ccEnclosing :: Provenance a
-    , _ccTypeCheckConfig :: PirTCConfig uni
+    , _ccEnclosing       :: Provenance a
+    , _ccTypeCheckConfig :: Maybe (PirTCConfig uni)
```
Comment (michaelpj) [inline plutus-core/plutus-ir/Language/PlutusIR/Compiler/Types.hs:57]: This now does two things, both telling us whether to typecheck and if so how to do it. Worth a comment at least.

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 2]
```diff
@@ -312,14 +283,15 @@ runCompiler opts expr = do
     let ctx = PIR.defaultCompilationCtx
               & set (PIR.ccOpts . PIR.coOptimize) (poOptimize opts)
               & set PIR.ccBuiltinMeanings PLC.getStringBuiltinMeanings
-              & set PIR.ccTypeCheckConfig (PIR.PirTCConfig stringBuiltinTCConfig PIR.YesEscape)
+              & set PIR.ccTypeCheckConfig (
+                  if poDoTypecheck opts
+                  then Just $ PIR.PirTCConfig stringBuiltinTCConfig PIR.YesEscape
+                  else Nothing)
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:289]: IMO this is an example of why the change you made is maybe not a good thing. Rather than being able to just set the typechecking config that we will need if we are going to typecheck, we also have to work in the decision about whether to typecheck or not.

At any rate, the usual rule about multiline expressions suggests putting this in a variable above.

File: plutus-tx-plugin/src/Language/PlutusTx/Compiler/Error.hs [hunk 3]
```diff
@@ -98,3 +102,6 @@ instance (PLC.GShow uni, PLC.Closed uni, uni `PLC.Everywhere` PLC.PrettyConst, P
         CompilationError e -> "Unexpected error during compilation, please report this to the Plutus team:" PP.<+> PP.pretty e
         UnsupportedError e -> "Unsupported feature:" PP.<+> PP.pretty e
         FreeVariableError e -> "Reference to a name which is not a local, a builtin, or an external INLINABLE function:" PP.<+> PP.pretty e
+        -- FIXME: transition into an sdoc solution for prettyprinting, to be compatible with ghc and its plugins
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Compiler/Error.hs:105]: I don't think we should do this. GHC's `SDoc` is a huge pain, and the sooner we can get rid of it the better.

File: plutus-tx-plugin/src/Language/PlutusTx/Compiler/Error.hs [hunk 4]
```diff
@@ -98,3 +102,6 @@ instance (PLC.GShow uni, PLC.Closed uni, uni `PLC.Everywhere` PLC.PrettyConst, P
...
+        InvalidMarkerError sdoc -> "Found invalid marker, not applied correctly" PP.<+> PP.pretty (GHC.showSDocUnsafe sdoc)
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Compiler/Error.hs:106]: Better to do the step of converting the `sdoc` into something else when we create the error term, and only have `Text` inside it.

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 5]
```diff
@@ -82,43 +122,32 @@ unconditionally which we pretty much are.
...
+mkPluginPass opts = \case
+    Just markerName -> GHC.CoreDoPluginPass "Core to PLC" $
+                  let ?markerName = markerName
+                      ?opts = opts
+                  in compileMod
+    Nothing -> GHC.CoreDoNothing -- nothing to do
+
+-- | The plugin works at haskell-module level granularity; the plugin
+-- looks at the module's the top-level bindings for plc markers and compile their inner core-expressions.
+compileMod :: (?markerName :: GHC.Name, ?opts :: PluginOptions)
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:136]: I really don't think we should use `ImplicitParams` unless we have a really good reason. If you really hate just passing the arguments to functions, why not just use `ReaderT`?

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 6]
```diff
@@ -63,7 +65,45 @@ data PluginOptions = PluginOptions {
...
+                       <$> parsePluginArgs args
+                       <*> getMarkerName
+          -- return the pipeline
+          pure $
+             simplPass
+             : pluginPass
+             : rest
+
+      -- <https://hackage.haskell.org/package/ghc-8.10.1/docs/GhcPlugins.html#v:thNameToGhcName> says
+      -- that we only have to call this once, since it is bound to an exact th symbol (using the 'quote syntax)
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:87]: I *think* this is okay, but it's subtle. `thNameToGhcName` can look up the name in the current `HscEnv`, which differs *per module*. So we might find the name or not depending whether we're in the context of a particular module when this is run. 

I don't know when `install` is run, but I'm suspicious. Whereas the previous version would only call it once we were definitely processing the module in question. So either we need a careful justification of this, or go back to the previous version which was more obviously correct (and that should anyway have had a note explaining this).

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs (hunk 5 above)
Comment (michaelpj) [reply plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:136]: You already put a monad transformer over `CoreM` below!

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 7]
```diff
@@ -139,86 +168,20 @@ with this, since you can't really specify a polymorphic type in a type applicati
...
-    in \case
+compileBind :: (?markerName :: GHC.Name, ?opts :: PluginOptions, ?famEnvs :: GHC.FamInstEnvs)
+            => GHC.CoreBind -> ECore PLC.DefaultUni GHC.CoreBind
+compileBind = \case
+    GHC.NonRec b rhs -> GHC.NonRec b <$> compileMarkers rhs
+    GHC.Rec bindsRhses -> GHC.Rec <$> for bindsRhses
+                         (\(b,rhs) -> (b,) <$> compileMarkers rhs)
+
+-- | Compiles all the core-expressions surrounded by plc in the given expression into PLC literals.
+compileMarkers :: (?markerName :: GHC.Name, ?opts :: PluginOptions, ?famEnvs :: GHC.FamInstEnvs)
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:182]: This is IMO a worse name than the previous one: we really do compile the *marked expressions*, not the markers!

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 8]
```diff
@@ -82,43 +122,32 @@ unconditionally which we pretty much are.
...
-    case maybeName of
-        -- nothing to do
-        Nothing   -> pure guts
-        Just name -> GHC.bindsOnlyPass (mapM $ compileMarkedExprsBind (opts, fam_envs) name) guts
+-- | The monad where the plugin runs in for each module.
+-- It is a core->core compiler monad, called CoreM, augmented with pure errors.
+type ECore uni = ExceptT (CompileError uni) GHC.CoreM
+runECore :: ECore PLC.DefaultUni a -> GHC.CoreM a
+runECore = runExceptT
+          >=> either (failCompilation . show . PP.pretty) pure
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:150]: IMO we should do the handling of `defer-errors` here. That would mean we have to specialise the runner so it's `ECore PLC.DefaultUni CoreExpr -> GHC.CoreM CoreExpr`, but that's fine.

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 9]
```diff
@@ -82,43 +122,32 @@ unconditionally which we pretty much are.
...
+      in GHC.bindsOnlyPass (runECore . mapM compileBind) guts
 
-    maybeName <- getMarkerName
-    case maybeName of
-        -- nothing to do
-        Nothing   -> pure guts
-        Just name -> GHC.bindsOnlyPass (mapM $ compileMarkedExprsBind (opts, fam_envs) name) guts
+-- | The monad where the plugin runs in for each module.
+-- It is a core->core compiler monad, called CoreM, augmented with pure errors.
+type ECore uni = ExceptT (CompileError uni) GHC.CoreM
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:147]: The usual convention is to call a concrete monad something M. The current one I guess is supposed to suggest it handles errors, but this is actually a bit restrictive since if e.g. we added a `ReaderT` it would become misleading. How about `PluginM`?

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 10]
```diff
@@ -230,74 +193,82 @@ compileMarkedExprs opts markerName =
...
+-- then the function returns a core expression that corresponds to a runtime error, to be injected back to the Haskell program.
+compileMarkedExpOrDefer :: (?opts :: PluginOptions, ?famEnvs :: GHC.FamInstEnvs)
+               => String -> GHC.Type -> GHC.CoreExpr -> ECore PLC.DefaultUni GHC.CoreExpr
+compileMarkedExpOrDefer locStr codeTy origE =
+    compileMarkedExpr locStr codeTy origE
+    & if poDeferErrors ?opts
+      then (`catchError` emitRuntimeError)
+      else id
+  where
+    emitRuntimeError :: CompileError PLC.DefaultUni -> ECore uni GHC.CoreExpr
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:226]: This whole thing could just be in `CoreM`, then you don't need all the `lift`s.

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 11]
```diff
@@ -230,74 +193,82 @@ compileMarkedExprs opts markerName =
...
+--
+-- If plugin-user has enabled 'defer-errors' and the PIR/PLC compilers have errored during typechecking/compilation,
+-- then the function returns a core expression that corresponds to a runtime error, to be injected back to the Haskell program.
+compileMarkedExpOrDefer :: (?opts :: PluginOptions, ?famEnvs :: GHC.FamInstEnvs)
+               => String -> GHC.Type -> GHC.CoreExpr -> ECore PLC.DefaultUni GHC.CoreExpr
+compileMarkedExpOrDefer locStr codeTy origE =
+    compileMarkedExpr locStr codeTy origE
+    & if poDeferErrors ?opts
+      then (`catchError` emitRuntimeError)
+      else id
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:224]: One reason I don't like this is that it makes the error handling look weirdly partial. In fact, any error is either used to fail compilation or put into the code, but here the two cases are separated.

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 12]
```diff
@@ -230,74 +193,82 @@ compileMarkedExprs opts markerName =
...
+          let shown = show $ PP.pretty (pruneContext (poContextLevel ?opts) e)
+          defUni <- lift . GHC.lookupTyCon =<< thNameToGhcNameOrFail ''PLC.DefaultUni
+          tcName <- thNameToGhcNameOrFail ''CompiledCode
+          tc <- lift $ GHC.lookupTyCon tcName
+          let args = [GHC.mkTyConTy defUni, codeTy]
+          pure $ GHC.mkRuntimeErrorApp GHC.rUNTIME_ERROR_ID (GHC.mkTyConApp tc args) shown
+
+-- | Actually invokes the Core to PIR to PLC compiler to compile an expression into a PLC literal.
+-- Compile the core expression that is surrounded by a 'plc' marker,
+-- and return the plc compiled ""bytecode" as a bytestring, to be injected back to the Haskell program
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:237]: It's not really "bytecode", it's just a serialied AST.

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 13]
```diff
@@ -230,74 +193,82 @@ compileMarkedExprs opts markerName =
...
+compileMarkedExpr :: (?opts :: PluginOptions, ?famEnvs :: GHC.FamInstEnvs)
+                => String -> GHC.Type -> GHC.CoreExpr -> ECore PLC.DefaultUni GHC.CoreExpr
+compileMarkedExpr locStr codeTy origE = do
+    -- module initialization
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:241]: ?

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 14]
```diff
@@ -230,74 +193,82 @@ compileMarkedExprs opts markerName =
...
     flags <- GHC.getDynFlags
-
-    -- We need to do this out here, since it has to run in CoreM
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:259]: This is a useful comment!

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 15]
```diff
@@ -230,74 +193,82 @@ compileMarkedExprs opts markerName =
...
+            ccBuiltinNameInfo = nameInfo,
+            ccBuiltinMeanings = PLC.getStringBuiltinMeanings,
+            ccScopes = initialScopeStack,
+            ccBlackholed = mempty
             }
-        initialState = CompileState {}
-    res <- runExceptT . runQuoteT . flip evalStateT initialState . flip runReaderT context $
-        withContextM 1 (sdToTxt $ "Compiling expr at" GHC.<+> GHC.text locStr) $ runCompiler opts origE
-    case res of
-        Left s ->
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:274]: I do like getting the handling of errors out of here, though :+1:

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 16]
```diff
@@ -336,3 +308,85 @@ runCompiler opts expr = do
 
     let uplcP = UPLC.eraseProgram plcP
     pure (pirP, uplcP)
+
+
+-- Helpers
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:313]: I'm not a huge fan of the "helpers" section. Usually you find functions by searching for their names if you're not sure where they are, and if anything you want them close to the place they're used.

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 17]
```diff
@@ -336,3 +308,85 @@ runCompiler opts expr = do
...
+    GHC.Tick _ e -> stripTicks e
+    e            -> e
+
+
+-- Helper to avoid doing too much construction of Core ourselves
+mkCompiledCode :: forall a . BS.ByteString -> BS.ByteString -> CompiledCode PLC.DefaultUni a
+mkCompiledCode plcBS pirBS = SerializedCode plcBS (Just pirBS)
+
+
+mkSimplPass :: GHC.DynFlags -> GHC.CoreToDo
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:370]: This one could be a local function next to `install`, I guess.

**Review by michaelpj — COMMENTED** [review]
I disagree about a few things :p

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs (hunk 11 above)
Comment (bezirg) [reply plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:224]: But the "meaning" of this is exactly the same as before this modification, correct?

Comment (michaelpj) [reply plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:224]: I think it probably does the right thing, but I had to think about it quite hard.

(Why do I care about how "obvious" it is that it does the right thing? a) It makes reading it easier in future, b) it makes it easier to change it without accidentally breaking things c) the more complicated the correctness reasoning is, the more likely you are to just make a mistake!)

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs (hunk 6 above)
Comment (bezirg) [reply plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:87]: I see your point now, and I understand this is not the optimal behaviour, since we may traverse looking for markers in a module for no reason, since 'plc is not in the local-module context. According to ghc doc: 

> thNameToGhcName :: Name -> CoreM (Maybe Name)Source#
Attempt to convert a Template Haskell name to one that GHC can understand. Original TH names such as those you get when you use the 'foo syntax will be translated to their equivalent GHC name exactly. Qualified or unqualified TH names will be dynamically bound to names in the module being compiled, if possible. Exact TH names will be bound to the name they represent, exactly.

The documentation is a bit confusing actually. I have a feeling that even if we move this `getMarkerName 'plc` to its older place, we do get the same suboptimal behaviour according to the docs. Since this is an "original th name" with `'foo` syntax, it means that even if we move it to be executed at the local-module level, failing to find the 'plc in the local context means that it will try in the global context as well. We perhaps have to manually construct a "Qualified or unqualified TH names" that  "will be **dynamically bound** to names in the module being compiled". Then if this manual th name symbol is not found locally, we do not have to traverse the whole binds looking for its occurences inside expressions.
This is all speculation  from just reading the ghc docs. I did not dig that deep in the ghc source code.

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs (hunk 10 above)
Comment (bezirg) [reply plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:226]: I see now that I need some lifts in there because I use the pure-error version: `thNameToGhcNameOrFail` function

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs (hunk 6 above)
Comment (bezirg) [reply plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:87]: I feel that using the 'foo syntax results to "exact GHC names", whereas perhaps using the (mkName "Module.plc") syntax results to dynamically bound names. But then dynamically bound names might have problems of their own.

Comment (bezirg) [reply plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs]: I moved it back at the location it used to be. I added a TODO: note about it.

File: plutus-core/plutus-ir/Language/PlutusIR/Compiler.hs [hunk 18]
```diff
@@ -59,25 +59,26 @@ floatTerm = runIfOpts letFloat
...
             pure $ LetFloat.floatTerm means t
 
--- | Perform typechecking of a PIR Term.
+-- | Typecheck a PIR Term iff the context demands it.
 -- Note: assumes globally unique names
-typeCheckTerm :: Compiling m e uni b => Term TyName Name uni (Provenance b) -> m ()
-typeCheckTerm t = do
-    tcconfig <- asks _ccTypeCheckConfig
-    void . runTypeCheckM tcconfig $ inferTypeM t
+mtypeCheckTerm :: Compiling m e uni b => Term TyName Name uni (Provenance b) -> m ()
```
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/plutus-ir/Language/PlutusIR/Compiler.hs]: ([msum](https://hackage.haskell.org/package/base-4.14.0.0/docs/Control-Monad.html#v:msum), [mfilter[(https://hackage.haskell.org/package/base-4.14.0.0/docs/Control-Monad.html#v:mfilter), agree with the point though)

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs (hunk 8 above)
Comment (bezirg) [reply plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs]: We could perhaps move this catchError to the "runExceptT" module-level, but
 it leads to uglier code and more difficult of handling other pure errors. I think I want to skip this (for now).

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 19]
```diff
@@ -228,88 +212,104 @@ compileMarkedExprs opts markerName =
...
+                 => GHC.Type -> CompileError uni fun -> PluginM uni fun GHC.CoreExpr
+emitRuntimeError codeTy e = do
+    opts <- asks pcOpts
+    let shown = show $ PP.pretty (pruneContext (poContextLevel opts) e)
+    tcName <- thNameToGhcNameOrFail ''CompiledCode
+    tc <- lift . lift $ GHC.lookupTyCon tcName
+    pure $ GHC.mkRuntimeErrorApp GHC.rUNTIME_ERROR_ID (GHC.mkTyConApp tc [codeTy]) shown
+
+-- | Compile the core expression that is surrounded by a 'plc' marker,
+-- and return the compiled plc AST as a serialized bytestring, to be injected back to the Haskell program.
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:259]:
```suggestion
-- and return a core expression which evaluates to the compiled plc AST as a serialized bytestring, to be injected back to the Haskell program.
```

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 20]
```diff
@@ -326,3 +326,61 @@ runCompiler opts expr = do
 
     let uplcP = UPLC.eraseProgram plcP
     pure (pirP, uplcP)
+
+-- | Get the 'GHC.Name' corresponding to the given 'TH.Name', or throw an exception if we can't get it.
```
Comment (michaelpj) [inline plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:330]:
```suggestion
-- | Get the 'GHC.Name' corresponding to the given 'TH.Name', or throw an error if we can't get it.
```

**Review by michaelpj — APPROVED** [review]
Just a few minor style things.

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs [hunk 21]
```diff
@@ -91,32 +127,55 @@ install args todo = do
...
-        Nothing   -> pure guts
-        Just name -> GHC.bindsOnlyPass (mapM $ compileMarkedExprsBind (opts, fam_envs) name) guts
+        Nothing -> pure guts
+        Just markerName ->
+            let pctx = PluginCtx { pcOpts = opts
+                                 , pcFamEnvs = (p_fam_env, GHC.mg_fam_inst_env guts)
+                                 , pcMarkerName = markerName
+                       }
+                -- start looking for plc calls from the top-level binds
+            in GHC.bindsOnlyPass (runPluginM pctx . mapM compileBind) guts
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:149]: I actually knew about that, but old habits hard.

File: plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs (hunk 20 above)
Comment (bezirg) [reply plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:330]: Once I read about the difference of exceptions VS errors, but in my head they always feel the same :stuck_out_tongue_closed_eyes:

Comment (michaelpj) [reply plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs:330]: It's loose, but in Haskell I think people do generally only use the word "exception" for the real runtime exceptions. "Error" is also vague, but at least it's not "exception" :laughing:
